### PR TITLE
Update the API layer

### PR DESCRIPTION
**Pull request by brian-c**
I made some big changes to the json-api-client lib to make it more predictable and less clumsy to work with.

The JSON-API lib no longer attempts to handle errors on its own, instead promises for failed requests are simply reject with the request. The rejected requests are no processed by the api client instance (/app/api/client) to work around the fact that different back end modules pass errors in different ways. A bad request now now consistently throws an `instanceof Error`, with a simple string `message` property (multi-line if needed) that's the best attempt at how to express the error.

These changes also lead to some refactoring of the auth module and the PromiseToSetState component mixin.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -31,9 +30,12 @@
     "uglify-js": "^2.4.16",
     "watchify": "^2.2.1"
   },
+  "peerDependencies": {
+    "coffeeify": "^1.0.0"
```
Comment (aweiksnar) [inline package.json]: I haven't been able to pick up coffeeify as a dependency on a fresh `npm i`, is yours locating it okay?

Comment (brian-c) [inline package.json]: Try `npm link` to link it globally, then `npm link panoptes` in another directory.

You don't need coffeeify unless you're requiring this as a lib in another project.

Comment (aweiksnar) [inline package.json]: I think the browserify transform in package.json is causing it to look for it. Running `npm start` is throwing `Error: Cannot find module 'coffeeify'`

Comment (aweiksnar) [inline package.json]: adding `  --ignore-transform coffeeify \ ` to the watchify command in bin/serve.sh seems to fix it

Comment (aweiksnar) [inline package.json]: Seems to be the case for all of the browserify calls in bin,  `test` & `build` we're the same

Comment (brian-c) [inline package.json]: Ugh, yeah, that makes sense. I just removed coffeeify and I get the same thing.

I'll push a fix in a minute.

**Comment by aweiksnar** [comment]
Cool, everything's running smoothly now. New error handling code looks really nice

**Pull request merged**